Initial

cryogenian opened the pull request:
There is no real tests, just couple checks that this works. I'll add them during work on `purescirpt-echarts` update. 

Fixes slamdata/slamdata#954

@garyb Please review

garyb commented:
I'm not at my desk right now so will review properly when I get home, but just from skimming on my phone this looks cool!

From a meta point of view, maybe we should break these up into separate repos though? Date and number formatting don't have much in common together, even though date formatting probably depends on number formatting.

garyb commented on src/Data/Formatter/DateTime.purs:
```diff
@@ -0,0 +1,426 @@
+  map f (Placeholder str a) = Placeholder str $ f a
+  map f End = End
+
+type Formatter = Mu FormatterF
```
What's the idea with using `Mu` here, rather than something like `newtype Formatter = List FormatterF`? I mean, it definitely works! But it wouldn't have occurred to me to use `Mu` for any non-tree structure.

cryogenian replied:
Actually I don't know why I chosed `Mu` rather then `List`. That was my first intent and it works, so I didn't even try `List`

garyb commented on src/Data/Formatter/DateTime.purs:
```diff
@@ -0,0 +1,426 @@
+  YearFull a →
+    (show $ fromEnum $ D.year d) <> cb a
+  YearTwoDigits a →
+    show ((fromEnum $ D.year d) - 1900) <> cb a
```
I think we'll need to do something cleverer here, as this only works for a small fraction of dates: `2016 - 1900 = 116`, `1066 - 1900 = -834`, etc. Admittedly the <1900 case is unlikely to be the kind of date formatted as YY, but dates in the 2000s almost certainly will be.

cryogenian replied:
Ok, I'll switch to something like `\x -> if ix > 30 then 1900 + x else 2000 + x`

garyb commented on src/Data/Formatter/DateTime.purs:
```diff
@@ -0,0 +1,426 @@
+
+formatDateTime ∷ String → DT.DateTime → Either String String
+formatDateTime pattern datetime =
+  parseFormatString pattern <#> flip format datetime
```
I kinda have a problem with this name, as both functions do indeed "format date/time" so it seems odd to prefix one with that. I don't really have a suggestion for a better name though, aside from my old fallback to `format'`...

cryogenian replied:
Yeah, I don't like these name too, but I didn't come up with something better. Any suggestions are really welcome.

garyb commented on src/Data/Formatter/DateTime.purs:
```diff
@@ -0,0 +1,426 @@
+  MonthTwoDigits a → do
+    ds ← some digit
+    let month = foldDigits ds
+    when (Arr.length ds /= 2 || month > 12 || month < 1) $ P.fail "Incorrect 2digit month"
```
I'd write "2digit" as "2-digit"

garyb commented on src/Data/Formatter/DateTime.purs:
```diff
@@ -0,0 +1,426 @@
+unformatAccumToDateTime a =
+  DT.DateTime
+    <$> (D.canonicalDate
+           <$> (maybe (Left "incorrect year") pure $ toEnum $ fromMaybe zero a.year)
```
The capitalisation of `Incorrect` here differs from some of the cases below.

Perhaps we could phrase it a little nicer too? "Could not parse year", etc. as "incorrect" could imply the value is somehow wrong (which is unknowable) rather than it just not matching the expected format.

garyb commented on src/Data/Formatter/DateTime.purs:
```diff
@@ -0,0 +1,426 @@
+  YearTwoDigits a → do
+    ds ← some digit
+    when (Arr.length ds /= 2) $ P.fail "Incorrect 2digit year"
+    lift $ modify _{year = Just $ foldDigits ds + 1900}
```
This case is even trickier than the formatting case. I'd expect the 2-digit year `16` to be `2016` rather than `1916`, so we should probably instate some switch where numbers over a certain value go into the 1900s, and numbers below go into the 2000s. How about 1970? Since that goes with the unix epoch also. So `70 -> 1970`, `69 -> 2069`, etc.

We could probably move it even nearer the current day, but the choice is pretty arbitrary.

garyb commented on src/Data/Formatter/DateTime.purs:
```diff
@@ -0,0 +1,426 @@
+    dow ← digit
+    when (dow > 7 || dow < 1) $ P.fail "Incorrect day of week"
+    lift $ modify _{dayOfWeek = Just dow}
+    cb a
```
This is a strange one in the unformat situation - do we need to store it at all? We basically ignore it when constructing the date, so keeping the info in the accum is unnecessary perhaps.

cryogenian replied:
Thanks, I've missed this stuff. Initially I thought about porting everything from `moment.js` and then validate string. E.g. If user tries to parse something like `12 -- 2` as `MM -- Q` he takes error `Month and Quater conflict`. But it's going to take too much time.

garyb commented on src/Data/Formatter/DateTime.purs:
```diff
@@ -0,0 +1,426 @@
+           <$> (maybe
+                  (Left "incorrect hour") pure
+                  $ toEnum
+                  $ (if a.meridiem == Just PM then (_ + 12) else id)
```
I think maybe we'll need to use an `Either` for `hour` or something, as you could have a (stupid) format like `HH a`, which if then was used on `23 pm` would get 35 as the hour here.

cryogenian replied:
Ah! I'll rather add this as special case, like if `meridiem == Just PM && hour > 12 then Left "Incorrect hour"`
Thanks!

garyb commented on src/Data/Formatter/Number.purs:
```diff
@@ -0,0 +1,222 @@
+  , format
+  , unformat
+  , formatNumber
+  , unformatNumber
```
Same thing as with the `formatDateTime` naming here - but again I have no good suggestion for the better alternative apart from priming.

garyb commented:
We could probably do with adding some documentation explaining the options for the format strings. The date ones I'm familiar with already, so had no problem there, but I'm having difficulty figuring out how the number format patterns actually work: does it matter where the comma appears, for instance, and what does "after" and "before" refer to with that? Places before/after the decimal point?

I'm not sure the numeric formats are actually working correctly either:

```
> FN.formatNumber "0.0" 123345.1235
Right { value0: '123,345.1' } -- is the comma supposed to be inserted?

> FN.formatNumber "0.0" (-123345.1235)
runtime exception! (unsafePartial'ed Maybe pattern failure)
```

cryogenian commented:
+ Fixed number format error 
+ Two digit year now checks if it's more then 69 and if so think that it's in 20th century otherwise in 21st. 
+ Changed to `2-digit`
+ Handling of pm + hour > 12

I think it's ok to have both formatters in one library because
+ They are both formatters 
+ They use same libs
+ If one wants formatting numbers we could be almost sure that he/she wants formatting datetimes and viceversa.

garyb commented on README.md:
```diff
@@ -1,2 +1,48 @@
+
+Formatter has following properties
++ Number of digits before dot
++ Numberr of digits after dot
```
Minor typo: "numberr" :smile:

garyb commented on README.md:
```diff
@@ -1,2 +1,48 @@
++ Numberr of digits after dot
++ Should sign be printed for positive numbers
++ Should thousands be separated by comma
++ Should output string has abbreviations (like `K` or `M`)
```
"has abbreviations" -> "have abbreviations"

garyb commented on README.md:
```diff
@@ -1,2 +1,48 @@
+
+Number of digits before dot is set by number of leading zeros:
++ `0000.0` will show 4 digits: `12 -> "0012.0"`, `1234 -> "1234.0"`
++ `00.0` will show only 2 digits : `1234 -> "1234.0, `1234 -> "1234.0"`
```
The quotes are a little off here: should be `1234 -> "1234.0"`,

cryogenian commented:
Updated readme

garyb commented on README.md:
```diff
@@ -1,2 +1,49 @@
+then use formatter to result of division of input number and that power.
++ `0a`: `1234567 -> "1M"`, `1 -> "1"`
+
+## Datetim formatters
```
"Datetim" -> "Date/Time"? :smile:

garyb commented:
When the tests pass this time (:wink:) :+1: from me. I'll hook up the dependency badges and stuff like that before making a release though, so if you just merge this I can make the actual release shortly after, if that's good with you?

cryogenian commented:
Yep, I was thinking about making something like `0.0.1` for this and then check what's actually needed by echarts, if there are any bugs etc and then switch to `0.1.0`. 

Thanks for review!

garyb commented:
Ah good point! Go ahead with the release then and I'll get the other stuff in place for the bumped version release.
